src/compiler/tsbuildPublic.ts

```diff
@@ -1568,42 +1733,68 @@ namespace ts {
         return actual;
     }
 
-    function updateOutputTimestampsWorker(state: SolutionBuilderState, proj: ParsedCommandLine, priorNewestUpdateTime: Date, verboseMessage: DiagnosticMessage, skipOutputs?: ESMap<Path, string>) {
-        if (proj.options.noEmit) return priorNewestUpdateTime;
+    function updateOutputTimestampsWorker(
+        state: SolutionBuilderState,
+        proj: ParsedCommandLine,
+        projectPath: ResolvedConfigFilePath,
+        verboseMessage: DiagnosticMessage,
+        skipOutputs?: ESMap<Path, string>
+    ) {
+        if (proj.options.noEmit) return;
+        let now: Date | undefined;
+        const buildInfoPath = getTsBuildInfoEmitOutputFilePath(proj.options);
+        if (buildInfoPath) {
+            if (!skipOutputs?.has(toPath(state, buildInfoPath))) {
+                if (!!state.options.verbose) reportStatus(state, verboseMessage, proj.options.configFilePath!);
+                state.host.setModifiedTime(buildInfoPath, now = getCurrentTime(state.host));
+                getBuildInfoCacheEntry(state, buildInfoPath, projectPath)!.modifiedTime = now;
+            }
+            state.outputTimeStamps.delete(projectPath);
+            return;
+        }
+
         const { host } = state;
         const outputs = getAllProjectOutputs(proj, !host.useCaseSensitiveFileNames());
+        const outputTimeStampMap = getOutputTimeStampMap(state, projectPath);
+        const modifiedOutputs = outputTimeStampMap ? new Set<Path>() : undefined;
         if (!skipOutputs || outputs.length !== skipOutputs.size) {
             let reportVerbose = !!state.options.verbose;
-            const now = host.now ? host.now() : new Date();
             for (const file of outputs) {
-                if (skipOutputs && skipOutputs.has(toPath(state, file))) {
-                    continue;
-                }
-
+                const path = toPath(state, file);
+                if (skipOutputs?.has(path)) continue;
                 if (reportVerbose) {
                     reportVerbose = false;
                     reportStatus(state, verboseMessage, proj.options.configFilePath!);
                 }
-
-                if (isDeclarationFileName(file)) {
-                    priorNewestUpdateTime = newer(priorNewestUpdateTime, getModifiedTime(host, file));
+                host.setModifiedTime(file, now ||= getCurrentTime(state.host));
+                if (outputTimeStampMap) {
+                    outputTimeStampMap.set(path, now);
+                    modifiedOutputs!.add(path);
                 }
-
-                host.setModifiedTime(file, now);
             }
         }
 
-        return priorNewestUpdateTime;
+        // Clear out timestamps not in output list any more
+        outputTimeStampMap?.forEach((_value, key) => {

```

Comment:
This is where we clean out any output time stamps for files taht are not in program any more..